File: lib-segment/src/main/java/websiteschema/mpsegment/hmm/Transition.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package websiteschema.mpsegment.hmm;

import java.io.IOException;
import websiteschema.mpsegment.util.ISerialize;
import websiteschema.mpsegment.util.SerializeHandler;

/**
 *
 * @author ray
 */
public class Transition implements ISerialize {

    Trie root = null;

    public Trie getRoot() {
        return root;
    }
    NodeRepository stateBank = new NodeRepository();
    TrieNodeSortor sortor = null;

    public void setStateBank(NodeRepository stateBank) {
        this.stateBank = stateBank;
    }

    public void setSortor(TrieNodeSortor sortor) {
        Trie.setTreeNodeSorter(sortor);
    }

    public Transition() {
        this.root = new Trie();
    }

    public Transition(Trie root, NodeRepository bank) {
        this.root = root;
        this.stateBank = bank;
    }

    public void setProb(int s1, int s2, double prob) {
        int[] ngram = new int[]{s1, s2};
        Trie node = root.insert(ngram);
        node.setProb(prob);
    }

    private double getProb(int[] ngram) {
        double ret;

        Trie node = root.searchNode(ngram);
        if (null != node) {
            ret = node.getProb();
        } else {
            ret = 1.0 / root.getCount();
        }

        return ret;
    }

    public double getCoProb(int[] c, int s) {
        int[] ngram = new int[c.length + 1];
        System.arraycopy(c, 0, ngram, 0, c.length);
        ngram[c.length] = s;

        return getProb(ngram, ngram.length);
    }

    public double getProb(int s1, int s2) {
        int[] ngram = new int[]{s1,s2};
        return getProb(ngram, 2);
    }

    public double getProb(int[] ngram, int n) {
        double ret = 0.00000001D;

        //bigram
        if (2 == n) {
            return getProb(ngram);
        }

        for (int i = n; i > 0; i--) {
            int[] igram = new int[i];
            for (int j = 1; j <= i; j++) {
                igram[i - j] = ngram[n - j];
            }
            ret += Flag.getInstance().labda(i - 1) * getProb(igram);
        }

        return ret;
    }

    private int[] getNodeArray(String[] ngram) {
        int[] array = new int[ngram.length];
        for(int i = 0; i < ngram.length; i++) {
            array[i] = stateBank.get(ngram[i]).getIndex();
        }
        return array;
    }
    
    private double getProb(String[] ngram) {
        double ret;

        Trie node = root.searchNode(getNodeArray(ngram));
        if (null != node) {
            ret = node.getProb();
        } else {
            ret = 1.0 / root.getCount();
        }

        return ret;
    }

    @Override
    public void save(SerializeHandler writeHandler) throws IOException {
        root.save(writeHandler);
        stateBank.save(writeHandler);
    }

    @Override
    public void load(SerializeHandler readHandler) throws IOException {
        root.load(readHandler);
        stateBank.load(readHandler);
    }
}

File: lib-segment-scala/src/main/scala/websiteschema/mpsegment/hmm/Transition.scala
package websiteschema.mpsegment.hmm

import websiteschema.mpsegment.util.ISerialize
import websiteschema.mpsegment.util.SerializeHandler

class Transition extends ITransition with ISerialize {

  var root: Trie = null

  def getRoot(): Trie = root

  var stateBank: NodeRepository = null

  def setStateBank(stateBank: NodeRepository) {
    this.stateBank = stateBank
  }

  def setProb(s1: Int, s2: Int, prob: Double) {
    val ngram = Array[Int](s1, s2)
    val node = root.insert(ngram)
    node.setProb(prob)
  }

  private def getProb(ngram: Array[Int]): Double = {
    var ret: Double = 0.0D

    val node = root.searchNode(ngram.toArray)
    if (null != node) {
      ret = node.getProb()
    } else {
      ret = 1.0 / root.getCount().toDouble
    }

    return ret
  }

  def getConditionProb(condition: Array[Int], state: Int): Double = {
    val ngram = new Array[Int](condition.length + 1)
    System.arraycopy(condition, 0, ngram, 0, condition.length)
    ngram(condition.length) = state

    return getProb(ngram, ngram.length)
  }

  def getConditionProb(condition: Int, state: Int): Double = {
    return getProb(Array[Int](condition, state), 2)
  }

  private def getProb(ngram: Array[Int], n: Int): Double = {
    var ret = 0.00000001D

    //bigram
    if (2 == n) {
      return getProb(ngram)
    }
    var i = n
    while (i > 0) {
      val igram = new Array[Int](i)
      for (j <- 1 to i) {
        igram(i - j) = ngram(n - j)
      }
      ret += Flag.labda(i - 1) * getProb(igram)
      i -= 1
    }

    return ret
  }

  override def save(writeHandler: SerializeHandler) {
    root.save(writeHandler)
    stateBank.save(writeHandler)
  }

  override def load(readHandler: SerializeHandler) {
    root.load(readHandler)
    stateBank.load(readHandler)
  }
}

object Transition {
  def apply() = {
    val transition = new Transition()
    transition.stateBank = new NodeRepository()
    transition.root = new Trie()
    transition
  }

  def apply(root: Trie, stateBank: NodeRepository) = {
    val transition = new Transition()
    transition.stateBank = stateBank
    transition.root = root
    transition
  }
}
